File: src-tauri/src/region_selection.rs
use crate::performance_limits::{PerformanceLimits, RegionSize};
use crate::region_selection_types::{
    LogicalPoint, PhysicalRegion, RegionSelection, RegionSelectionIssue, RegionSelectionIssueCode,
    RegionSelectionLimits, RegionSelectionRequest, RegionSelectionResult,
};

const MIN_REGION_WIDTH: i32 = 24;
const MIN_REGION_HEIGHT: i32 = 24;

impl Default for RegionSelectionLimits {
    fn default() -> Self {
        let performance_limits = PerformanceLimits::default();

        Self {
            minimum_region: RegionSize {
                width: MIN_REGION_WIDTH,
                height: MIN_REGION_HEIGHT,
            },
            recommended_region: performance_limits.recommended_region,
            max_region: performance_limits.max_region,
        }
    }
}

pub fn select_region(request: RegionSelectionRequest) -> RegionSelectionResult {
    validate_geometry(&request)?;

    let limits = RegionSelectionLimits::default();
    let region = map_logical_selection_to_physical(&request)?;

    validate_minimum_region(&region, &limits)?;
    validate_hard_limits(&region, &limits)?;

    Ok(RegionSelection {
        warnings: recommended_warnings(&region, &limits),
        region,
        limits,
    })
}

fn validate_geometry(request: &RegionSelectionRequest) -> Result<(), RegionSelectionIssue> {
    let points = [
        request.monitor.logical_origin,
        LogicalPoint {
            x: request.monitor.physical_origin.x as f64,
            y: request.monitor.physical_origin.y as f64,
        },
        request.start,
        request.end,
    ];

    if points
        .iter()
        .any(|point| !point.x.is_finite() || !point.y.is_finite())
    {
        return Err(issue(RegionSelectionIssueCode::InvalidCoordinate, 0.0, 0.0));
    }

    if !request.monitor.scale_factor.is_finite() || request.monitor.scale_factor <= 0.0 {
        return Err(issue(
            RegionSelectionIssueCode::InvalidScaleFactor,
            0.0,
            finite_or_zero(request.monitor.scale_factor),
        ));
    }

    Ok(())
}

fn map_logical_selection_to_physical(
    request: &RegionSelectionRequest,
) -> Result<PhysicalRegion, RegionSelectionIssue> {
    let left = request.start.x.min(request.end.x);
    let right = request.start.x.max(request.end.x);
    let top = request.start.y.min(request.end.y);
    let bottom = request.start.y.max(request.end.y);

    let (x, width) = map_axis(
        left,
        right,
        request.monitor.logical_origin.x,
        request.monitor.physical_origin.x,
        request.monitor.scale_factor,
    )?;
    let (y, height) = map_axis(
        top,
        bottom,
        request.monitor.logical_origin.y,
        request.monitor.physical_origin.y,
        request.monitor.scale_factor,
    )?;

    Ok(PhysicalRegion {
        monitor_id: request.monitor.id.clone(),
        x,
        y,
        width,
        height,
    })
}

fn map_axis(
    min: f64,
    max: f64,
    logical_origin: f64,
    physical_origin: i32,
    scale: f64,
) -> Result<(i32, i32), RegionSelectionIssue> {
    let start = checked_i32(((min - logical_origin) * scale).floor() + physical_origin as f64)?;
    let end = checked_i32(((max - logical_origin) * scale).ceil() + physical_origin as f64)?;
    let width = i64::from(end) - i64::from(start);

    if width < 0 || width > i64::from(i32::MAX) {
        return Err(issue(
            RegionSelectionIssueCode::RegionCoordinateOutOfRange,
            i32::MAX as f64,
            width as f64,
        ));
    }

    Ok((start, width as i32))
}

fn checked_i32(value: f64) -> Result<i32, RegionSelectionIssue> {
    if !value.is_finite() || value < i32::MIN as f64 || value > i32::MAX as f64 {
        return Err(issue(
            RegionSelectionIssueCode::RegionCoordinateOutOfRange,
            i32::MAX as f64,
            value,
        ));
    }

    Ok(value as i32)
}

fn validate_minimum_region(
    region: &PhysicalRegion,
    limits: &RegionSelectionLimits,
) -> Result<(), RegionSelectionIssue> {
    if region.width < limits.minimum_region.width {
        return Err(issue(
            RegionSelectionIssueCode::RegionTooNarrow,
            limits.minimum_region.width as f64,
            region.width as f64,
        ));
    }

    if region.height < limits.minimum_region.height {
        return Err(issue(
            RegionSelectionIssueCode::RegionTooShort,
            limits.minimum_region.height as f64,
            region.height as f64,
        ));
    }

    Ok(())
}

fn validate_hard_limits(
    region: &PhysicalRegion,
    limits: &RegionSelectionLimits,
) -> Result<(), RegionSelectionIssue> {
    if region.width > limits.max_region.width {
        return Err(issue(
            RegionSelectionIssueCode::RegionWidthTooLarge,
            limits.max_region.width as f64,
            region.width as f64,
        ));
    }

    if region.height > limits.max_region.height {
        return Err(issue(
            RegionSelectionIssueCode::RegionHeightTooLarge,
            limits.max_region.height as f64,
            region.height as f64,
        ));
    }

    Ok(())
}

fn recommended_warnings(
    region: &PhysicalRegion,
    limits: &RegionSelectionLimits,
) -> Vec<RegionSelectionIssue> {
    let mut warnings = Vec::with_capacity(2);

    if region.width > limits.recommended_region.width {
        warnings.push(issue(
            RegionSelectionIssueCode::RegionWidthAboveRecommended,
            limits.recommended_region.width as f64,
            region.width as f64,
        ));
    }

    if region.height > limits.recommended_region.height {
        warnings.push(issue(
            RegionSelectionIssueCode::RegionHeightAboveRecommended,
            limits.recommended_region.height as f64,
            region.height as f64,
        ));
    }

    warnings
}

fn issue(code: RegionSelectionIssueCode, limit: f64, actual: f64) -> RegionSelectionIssue {
    RegionSelectionIssue {
        code,
        limit: finite_or_zero(limit),
        actual: finite_or_zero(actual),
        message: message_for(code).to_string(),
    }
}

fn finite_or_zero(value: f64) -> f64 {
    if value.is_finite() {
        value
    } else {
        0.0
    }
}

fn message_for(code: RegionSelectionIssueCode) -> &'static str {
    match code {
        RegionSelectionIssueCode::InvalidCoordinate => "Selection coordinates must be finite.",
        RegionSelectionIssueCode::InvalidScaleFactor => {
            "Monitor scale factor must be greater than zero."
        }
        RegionSelectionIssueCode::RegionTooNarrow => "Selected region is too narrow.",
        RegionSelectionIssueCode::RegionTooShort => "Selected region is too short.",
        RegionSelectionIssueCode::RegionCoordinateOutOfRange => {
            "Selected region is outside the supported coordinate range."
        }
        RegionSelectionIssueCode::RegionWidthAboveRecommended => {
            "Selected region is wider than recommended."
        }
        RegionSelectionIssueCode::RegionHeightAboveRecommended => {
            "Selected region is taller than recommended."
        }
        RegionSelectionIssueCode::RegionWidthTooLarge => "Selected region is wider than allowed.",
        RegionSelectionIssueCode::RegionHeightTooLarge => "Selected region is taller than allowed.",
    }
}

File: src/features/region-selector/regionSelection.ts
import {
  REGION_SELECTION_LIMITS,
  type LogicalPoint,
  type LogicalSize,
  type PhysicalPoint,
  type PhysicalRegion,
  type RegionSelectionIssue,
  type RegionSelectionLimits,
  type RegionSelectionRequest,
  type RegionSelectionResult
} from "./regionSelectionTypes";
import { isRegionSelectionIssue, issue } from "./regionSelectionIssues";
import { mapLogicalSelectionToPhysical } from "./regionSelectionMapper";

export {
  REGION_SELECTION_LIMITS,
  type LogicalPoint,
  type LogicalSize,
  type MonitorGeometry,
  type PhysicalPoint,
  type PhysicalRegion,
  type RegionSelection,
  type RegionSelectionIssue,
  type RegionSelectionIssueCode,
  type RegionSelectionLimits,
  type RegionSelectionRequest,
  type RegionSelectionResult
} from "./regionSelectionTypes";

export function selectRegion(
  request: RegionSelectionRequest,
  limits: RegionSelectionLimits = REGION_SELECTION_LIMITS
): RegionSelectionResult {
  const geometryError = validateGeometry(request);

  if (geometryError) {
    return { ok: false, error: geometryError };
  }

  const mappedRegion = mapLogicalSelectionToPhysical(request);

  if (isRegionSelectionIssue(mappedRegion)) {
    return { ok: false, error: mappedRegion };
  }

  const region = mappedRegion;
  const validationError =
    validateMinimumRegion(region, limits) ?? validateHardLimits(region, limits);

  if (validationError) {
    return { ok: false, error: validationError };
  }

  return {
    ok: true,
    selection: {
      region,
      warnings: recommendedWarnings(region, limits),
      limits
    }
  };
}

function validateGeometry(
  request: RegionSelectionRequest
): RegionSelectionIssue | null {
  const points = [
    request.monitor.logicalOrigin,
    request.monitor.physicalOrigin,
    request.start,
    request.end
  ];

  if (points.some((point) => !isFinitePoint(point))) {
    return issue("invalidCoordinate", 0, 0);
  }

  if (
    !Number.isFinite(request.monitor.scaleFactor) ||
    request.monitor.scaleFactor <= 0
  ) {
    return issue("invalidScaleFactor", 0, request.monitor.scaleFactor);
  }

  if (!isFinitePositiveSize(request.monitor.logicalSize)) {
    return issue("invalidCoordinate", 0, 0);
  }

  if (!selectionIsInsideMonitor(request)) {
    return issue("selectionOutsideMonitor", 0, 0);
  }

  return null;
}

function selectionIsInsideMonitor(request: RegionSelectionRequest): boolean {
  const left = request.monitor.logicalOrigin.x;
  const top = request.monitor.logicalOrigin.y;
  const right = left + request.monitor.logicalSize.width;
  const bottom = top + request.monitor.logicalSize.height;

  return [request.start, request.end].every(
    (point) =>
      point.x >= left &&
      point.x <= right &&
      point.y >= top &&
      point.y <= bottom
  );
}

function validateMinimumRegion(
  region: PhysicalRegion,
  limits: RegionSelectionLimits
): RegionSelectionIssue | null {
  if (region.width < limits.minimumRegion.width) {
    return issue("regionTooNarrow", limits.minimumRegion.width, region.width);
  }

  if (region.height < limits.minimumRegion.height) {
    return issue("regionTooShort", limits.minimumRegion.height, region.height);
  }

  return null;
}

function validateHardLimits(
  region: PhysicalRegion,
  limits: RegionSelectionLimits
): RegionSelectionIssue | null {
  if (region.width > limits.maxRegion.width) {
    return issue("regionWidthTooLarge", limits.maxRegion.width, region.width);
  }

  if (region.height > limits.maxRegion.height) {
    return issue("regionHeightTooLarge", limits.maxRegion.height, region.height);
  }

  return null;
}

function recommendedWarnings(
  region: PhysicalRegion,
  limits: RegionSelectionLimits
): RegionSelectionIssue[] {
  const warnings: RegionSelectionIssue[] = [];

  if (region.width > limits.recommendedRegion.width) {
    warnings.push(
      issue(
        "regionWidthAboveRecommended",
        limits.recommendedRegion.width,
        region.width
      )
    );
  }

  if (region.height > limits.recommendedRegion.height) {
    warnings.push(
      issue(
        "regionHeightAboveRecommended",
        limits.recommendedRegion.height,
        region.height
      )
    );
  }

  return warnings;
}

function isFinitePoint(point: LogicalPoint | PhysicalPoint): boolean {
  return Number.isFinite(point.x) && Number.isFinite(point.y);
}

function isFinitePositiveSize(size: LogicalSize): boolean {
  return (
    Number.isFinite(size.width) &&
    Number.isFinite(size.height) &&
    size.width > 0 &&
    size.height > 0
  );
}
